### PR TITLE
fix: Stop generator from getting stuck on circular relations.

### PR DESCRIPTION
**Pull request by SandPod**
Issue: https://github.com/serverpod/serverpod/issues/1267

We aimed to sort our dependencies so that any table that had a relation to another table would be defined after the table was created in our sql scripts.

This results in an infinite loop if we have a circular relation.

This PR changes our script creation behavior so that we:
1. Only sort on table names.
2. Always define our relations last in the file (this way the tables required are guaranteed to exists).

This fix is applied to both the legacy pgsql generator and the new migration feature.

The solution doesn't need to be applied to the alter table, since  we don't handle adding a foreign relation to the migration scripts yet.

Closes: https://github.com/serverpod/serverpod/issues/1267

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

This will generated sql files with the same functionality but in a different way.

### CODE REVIEW

**Comment by SandPod** [comment]
Might want to apply my quick fix here to cause a timeout when the generator gets stuck in an infinite loop: https://github.com/dart-lang/test/issues/2077#issuecomment-1673469007

File: tools/serverpod_cli/test/generator/psql/regression_test.dart [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    var codeMap = generator.generateSerializableEntitiesCode(
+      entities: entities,
+      config: config,
+    );
+
+    test('then expected file is created.', () async {
+      expect(codeMap[expectedFileName], isNotNull);
+    });
+
+    test('then file content matches regression data.', () async {
```
Comment (SandPod) [inline tools/serverpod_cli/test/generator/psql/regression_test.dart:57]: Comment: This test is not strictly necessary, it was just a nice way to ensure that nothing disappeared from the output when changing the code.

File: tools/serverpod_cli/lib/src/generator/psql/legacy_pgsql_generator.dart [hunk 2]
```diff
@@ -134,19 +90,34 @@ class LegacyPgsqlCodeGenerator extends CodeGenerator {
...
     }
 
+    return out;
+  }
+
+  String _generateForeignKeys(ClassDefinition classInfo) {
+    var out = '';
+
+    var relationFields = classInfo.fields.where((field) =>
+        field.relation != null && field.relation is ForeignRelationDefinition);
```
Comment (Isakdl) [inline tools/serverpod_cli/lib/src/generator/psql/legacy_pgsql_generator.dart:100]: Suggestion: convert to block function for formatting.

**Review by Isakdl — APPROVED** [review]
LGTM :) one NIT comment.

**Pull request merged**